Store a batch operation's  actor info in ES/OS secondary storage

korthout opened the pull request:
## Description

<!-- Describe the goal and purpose of this PR. -->

Adds the `actorType` and `actorId` to the batch operation entity stored in the ES/OS secondary storage.

The info is only added on batch operations created from 8.9 onwards. Older batch operations will have `null` for both fields.

## Checklist

<!--- Please delete options that are not relevant. Boxes should be checked by reviewer. -->
- [ ] Enable backports when necessary (fex. [for bug fixes](https://github.com/camunda/camunda/blob/main/CONTRIBUTING.md#backporting-changes) or [for CI changes](https://github.com/camunda/camunda/wiki/CI-&-Automation#when-to-backport-ci-changes)).

## Related issues

closes #42065

korthout committed:
feat: add actor info to batch operation entity

We'll need to store the actor type and actor id in the secondary
storage.

The data structure is copied from AuditLogEntity which also has an actor
type and actor id. While these are the same, I've chosen to use a
separate enum for the types as it's fine to keep these separate and the
actor type enum for the audit log entity already had a specific name
including audit log.

Note that these values will be filled starting from 8.9 only. I'll add a
check using the brokerVersion of the record to ensure that this will
hold true, once we get to the handler changes.

For older batch operations, we'll simply use `null` because we no longer
know who started it.

I've considered using the username as a fallback, but it won't be used
in V2. It's a leftover from the Operate V1 backend.

korthout committed:
feat: update BO entity with actor info

We should only fill the batch operation entity with actor info starting
from 8.9.0 onwards, as these fields are annotated as
@SinceVersion(8.9.0). We can check the broker version of the record for
this, to ensure we don't accidentally store it for a batch operation
created on 8.8 but exported after updating to 8.9.

The implementation is straightforward, with usage of SemanticVersion to
allow easily comparing the versions (this class is fast and uses
caching), so it's fine to use in an exporter.

We can simply return a record with the extracted values, or null values
so it can be directly used to update the entity.

I've added tests that shows some of the basic usage patterns:
- before 8.9
- after 8.9 with client id
- after 8.9 with username

korthout committed:
feat: add actor info to batch operation templates

Both should be indexed so keyword.

mustafadagher reviewed (approved):
Thanks @korthout ! 🚀  
I just left some thoughts about the `ActorInfoExtractor`.

korthout commented on webapps-schema/src/main/resources/schema/opensearch/create/template/operate-batch-operation.json:
```diff
@@ -31,6 +31,12 @@
+      },
       "state": {
         "type": "keyword"
       },
```
Yes, spotless can do indentation, see [generic steps](https://github.com/diffplug/spotless/blob/main/plugin-maven/README.md#generic-steps). 

I'll open a separate PR for this.

korthout replied:
I've opened https://github.com/camunda/camunda/pull/42545

korthout committed:
refactor: simplify AuditActorInfo construction

While it can be useful to build AuditLogInfo with the .of(Record)
method, we don't always need all the audit log info. It can be useful to
access the AuditLogActor independently.

korthout committed:
refactor: simplify using early return

This code could be simplified by only accessing each Authorization once
and storing it in a local variable.

Additionally, it could be simplified by applying early returns, removing
the redundant else blocks.

korthout committed:
refactor: re-use AuditLogActor

This removes the need to extract the information from the authorizations
twice.

I'm keeping the actor type in the webapps schema separated though. A
batch operation has a batch operation actor type, not an audit log
actor.

mustafadagher reviewed (approved):
LGTM! Thanks @korthout !
